[Tree-widget]: initial version for Tree widget consuming kiwi

MartynasStrazdas opened the pull request:
Initial preview of kiwi consumption in Tree widget

Demo:
https://github.com/user-attachments/assets/4cc0d50a-6c18-4709-84e5-5a8e58f72eed

There are some known issues, such as button alignment and item selection on expander click that will be fixed later.

grigasp commented:
I created a `tree-widget/next` branch for until the tree widget is ready for a full major release. Please re-target your PR to that.

grigasp reviewed (commented):
Why did lockfiles of unrelated packages changed? Can we revert those changes?

MartynasStrazdas commented:
> Why did lockfiles of unrelated packages changed? Can we revert those changes?

We added pnpmfile to handle itwinUI v5 peer dependency, still looking if there's a workaround of it not being in root.

MartynasStrazdas commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/common/components/Tree.tsx:
```diff
@@ -2,12 +2,14 @@
  * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
  * See LICENSE.md in the project root for license terms and full copyright notice.
  *--------------------------------------------------------------------------------------------*/
+/* eslint-disable deprecation/deprecation */
```
Left over from unified selection.

saskliutas commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/common/components/TreeNodeVisibilityButton.tsx:
```diff
@@ -0,0 +1,83 @@
+    }
+  };
+  return (
+    <TooltipWrapper content={checkboxState.tooltip}>
```
Tooltip should not be necessary if you are using `IconButton` with label

MartynasStrazdas replied:
This is a itwinUI issue, label is required and generates a tooltip, our tooltip loads a bit later which causes it to be undefined for a bit, which means we can't use it as a label. We probably want a label and a separate tooltip which.

MartynasStrazdas replied:
changed

MartynasStrazdas commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/common/components/TreeRenderer.tsx:
```diff
@@ -2,124 +2,30 @@
-    /** Callback that is invoked when node is double clicked. */
-    onNodeDoubleClick?: (node: PresentationHierarchyNode, isSelected: boolean) => void;
-  };
+export type TreeRendererProps = React.ComponentPropsWithoutRef<typeof PresentationTree> & { checkboxProps?: TreeItemVisibilityButtonProps };
```
Its for visibility button, adjusted name.

MartynasStrazdas commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/common/components/TreeNodeVisibilityButton.tsx:
```diff
@@ -0,0 +1,83 @@
+ * Data structure that describes tree node checkbox state.
+ * @beta
+ */
+interface TreeNodeCheckboxState {
```
renamed

MartynasStrazdas commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/common/components/Tree.tsx:
```diff
@@ -79,9 +89,9 @@ export function Tree({ getSchemaContext, hierarchyLevelSizeLimit, selectionStora
   return (
     <SchemaMetadataContextProvider imodel={props.imodel} schemaContextProvider={getSchemaContext}>
-      <UnifiedSelectionProvider storage={selectionStorage}>
+      <UnifiedSelectionContextProvider storage={selectionStorage}>
```
Moved.

saskliutas reviewed (changes requested):
Looks good. I would remove all the commented out code and props for now and track those removed features in an issue.

MartynasStrazdas commented on packages/itwin/tree-widget/src/tree-widget-react/components/TreeSelector.tsx:
```diff
@@ -18,7 +17,7 @@ export interface TreeContentDefinition {
   label: string;
   render: (props: SelectableTreeRenderProps) => React.ReactNode;
-  startIcon?: React.ReactNode;
+  startIcon?: string;
```
Currently button does not support `startIcon`, which means we need to use <Icon /> which accepts only href, probably can add | React.ReactElement, but there would be problems like the icon size unpredictability.

grigasp replied:
> Currently button does not support `startIcon`, which means we need to use which accepts only href

If we take `ReactNode`, couldn't consumers just pass us this: `<Icon href="path to icon" />`?

> there would be problems like the icon size unpredictability

Yeah, but if we say in docs what's the max allowed size, then it's not our problem.

I think it can stay like this for now, but please add a TODO comment to switch it back to `ReactNode` before the official major release.

MartynasStrazdas replied:
Reverted it back to react node, this is an internal component which is being used by selectableTree which takes in startIcon as reactNode, we should also use not have startIcon type be dublicated (added a TODO will look over more interfaces to see if we dublicate stuff)

MartynasStrazdas commented:
> > Why did lockfiles of unrelated packages changed? Can we revert those changes?
> 
> We added pnpmfile to handle itwinUI v5 peer dependency, still looking if there's a workaround of it not being in root.

Fixed

saskliutas commented on packages/itwin/tree-widget/src/test/EmptyTest.test.tsx:
Why is this needed? Maybe just remove `test` script from tree-widget-react package.json?

MartynasStrazdas replied:
Wouldn't we then need to remove it from pipeline and other places, would rather just keep it.

saskliutas replied:
It runs through `lage` in pipeline so it should be enough to remove `cover`/`test` scripts from package.json

MartynasStrazdas replied:
removed.

saskliutas commented:
@grigasp I think we should disable e2e tests for this branch. At least until we reach some stability. Because I think current changes to test-viewer updates property-grid UI too (colors, font, spacing, etc).

grigasp commented:
> @grigasp I think we should disable e2e tests for this branch. At least until we reach some stability. Because I think current changes to test-viewer updates property-grid UI too (colors, font, spacing, etc).

I think made them non-required check for PR, so they shouldn't prevent PR from being merged. But yeah, I agree it makes sense to just disable them for now.
